circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/navstack/SaveableNavStack.kt

```diff
@@ -0,0 +1,346 @@
+    return stateStore.keys.toList()
+  }
+
+  override fun removeState(screen: Screen): Boolean {
+    return stateStore.remove(screen) != null
+  }
+
+  override fun containsRecord(record: Record, includeSaved: Boolean): Boolean {
+    // If it's in the main entry list, return true
+    if (record in entryList) return true
+
+    if (includeSaved && stateStore.isNotEmpty()) {
+      // If we're checking our saved snapshots too, iterate through them and check
+      for (snapshot in stateStore.values) {
+        if (record in snapshot.entries) return true
+      }
+    }
+    return false
+  }
+
+  override fun isRecordReachable(key: String, depth: Int, includeSaved: Boolean): Boolean {
+    if (depth < 0) return false
+    // Check in the current entry list
+    if (isRecordReachable(key, depth, currentIndex, entryList)) {
+      return true
+    }
+    // If includeSaved, check saved stack states too
+    if (includeSaved && stateStore.isNotEmpty()) {
+      for (snapshot in stateStore.values) {
+        if (isRecordReachable(key, depth, snapshot.currentIndex, snapshot.entries)) {
+          return true
+        }
+      }
+    }
+    return true
+  }
+
+  private fun isRecordReachable(
+    key: String,
+    depth: Int,
+    index: Int,
+    records: List<Record>,
+  ): Boolean {
+    val min = maxOf(0, index - depth)
+    for (i in min until index) {
+      if (records[i].key == key) return true
+    }
+    val max = minOf(index + depth, records.size)
+    for (i in index until max) {
+      if (records[i].key == key) return true
+    }
+    return false
+  }
+
+  public data class Record(
+    override val screen: Screen,
+    val args: Map<String, Any?> = emptyMap(),
+    @OptIn(ExperimentalUuidApi::class) override val key: String = Uuid.random().toString(),
+  ) : NavStack.Record {
+
+    internal companion object {
+      val Saver: Saver<Record, Any> =
+        mapSaver(
+          save = { value ->
+            buildMap {
+              put("screen", value.screen)
+              put("args", value.args)
+              put("key", value.key)
+            }
+          },
+          restore = { map ->
+            @Suppress("UNCHECKED_CAST")
+            Record(
+              screen = map["screen"] as Screen,
+              args = map["args"] as Map<String, Any?>,
+              key = map["key"] as String,
+            )
+          },
+        )
+    }
+  }
+
+  internal data class SaveableNavStackList(val entries: List<Record>, val currentIndex: Int) :
+    NavStackList<Record> {
+
+    override val top: Record
+      get() = entries.first()
+
+    override val active: Record
+      get() = entries[currentIndex]
+
+    override val root: Record
+      get() = entries.last()
+
+    override val forwardItems: Iterable<Record>
+      get() = entries.subList(0, currentIndex).asReversed()
+
+    override val backwardItems: Iterable<Record>
+      get() = entries.subList(currentIndex + 1, entries.size)
```

Comment:
These have no checks for the `currentIndex`
being within the limits of the `entries` size,
is that a concern here?

Eg. you could pass
```
entries = [A, B, C],
currentIndex = 3
```
and this would crash with `IllegalArgumentException`